test/test_incorrect_dispatch_key_ops.py

```diff
@@ -0,0 +1,112 @@
+import os
+import re
+from typing import List
+
+from tools.codegen.gen import parse_native_yaml, get_grouped_native_functions
+from tools.codegen.model import DispatchKey, NativeFunction, NativeFunctionsGroup
+from torch.testing._internal.common_utils import TestCase, run_tests
+
+# The test below is a quick-and-dirty check to ensure that people mark operators in native_functions.yaml
+# with valid dispatch keys.
+# In particular, it's INCORRECT to implement an operator using DispatchStub, but mark the operator
+# with the CompositExplicitAutograd key in native_functions.yaml.
+# This is bad for two reasons:
+# - CompositeExplicitAutograd implies that the op is valid on all backends, but it isn't;
+#   it will break for external backends
+# - The cuda kernel registered to DispatchStub expects a device guard to be set; but the codegen
+#   doesn't create one for ComositeExplicitAutograd entries, because it expects Composite ops
+#   to call into sub-ops that themselves are registered to the CUDA key, which will have
+#   their own device guards.
+# See an example bug from this at https://github.com/pytorch/pytorch/issues/60892
+
+# The test written below is pretty fragile; it parses the aten/ folder to look for all
+# dispatch stub kernels, and maps them to native_functions.yaml ops through naming conventions.
+# I had to add a list of ignored ops here to avoid false positives.
+# For example, `at::native::cumprod` is a (valid) CompositeExplicitAutograd kernel, which is implemented
+# by calling directly into `at::native::_cumprod` - which has a CPU/CUDA kernel that calls `cumprod_stub()`,
+# which is registered to DispatchStub.
+# This is a kind of weird pattern, but it's valid, and hard to avoid catching in this test.
+#
+# If you create a future op that's implemented similarly and you're sure it's a false positive,
+# add it to this list.
+IGNORED_OPS = [
+    "cumprod",
+    "cumsum",
+    "logcumsumexp",
+]
+
+class TestIncorrectDispatchKeyOps(TestCase):
+
+    def find_dispatch_stub_kernels(self) -> List[str]:
+        curr_path = os.path.dirname(os.path.realpath(__file__))
+        root_dir = os.path.join(curr_path, '../aten/src/ATen/')
+        regex = re.compile("DEFINE_DISPATCH\((.*)_stub\)")
+        f_path = os.path.join(curr_path, '../aten/src/ATen/native/BinaryOps.cpp')
+
+        matches = []
+        for root, dirs, files in os.walk(root_dir):
```

Comment:
Good hygiene for os.walk is to ignore private directories (directories staring with period). In this case it doesn't matter because you start traversing at aten/src/ATen, but failure to filter those directories out is a common cause of os.walk calls taking a really looooong time to run. Actually, even better would probably be to use `git ls-files` but that would require PyTorch to be a Git checkout (which is not necessarily true when people are running tests)